### PR TITLE
#2543 add isEditable helpers

### LINKED ISSUE #2543: Update prescription tabs to enforce read-only constraints for non-local patients
## Is your feature request related to a problem? Please describe.

Patient lookup functionality will enable users to prescribe to non-local patients. To ensure data integrity, users will not be able to edit these patients or associated data (i.e. insurance policies). 

## Describe the solution you'd like

Update the prescription wizard to enforce the following constraints on non-local patients:

- Patient details are read-only.
- Patient insurance policies are read-only.

## Implementation

- Add a getter to `Name`, e.g. `isLocal` which returns if `supplyingStoreId` is current store.
- If prescribing for a patient where `isLocal` is `false`, disable patient, insurance editing.

## Describe alternatives you've considered

N/A. 

## Additional context

See epic #2446.

### PR DESCRIPTION
**Pull request by wlthomson**
Fixes #2543.

Branches off #2674.

## Change summary

Looks like most of the functionality needed is already implemented! Just added a few helpers to make things a bit more explicit.

## Testing

Steps to reproduce or otherwise test the changes of this PR:

- [ ] Patients which are added using the API cannot be edited locally.
- [ ] Insurance polices which are added using the API cannot be edited locally.
- [ ] Prescribers which are added using the API cannot be edited locally.

### Related areas to think about

N/A.

### CODE REVIEW

File: src/selectors/patient.js [hunk 1]
```diff
@@ -50,7 +50,6 @@ export const selectPatientModalOpen = ({ patient }) => {
 
 export const selectCanEditPatient = ({ patient }) => {
   const { currentPatient } = patient;
-  const { thisStoresPatient } = currentPatient ?? {};
-
-  return thisStoresPatient ?? true;
+  const { isEditable = true } = currentPatient ?? {};
```
Comment (joshxg) [inline src/selectors/patient.js:53]: possibly better to have this as false as the default

Comment (wlthomson) [reply src/selectors/patient.js]: Good point. 6d37591.

**Comment by wlthomson** [comment]
@joshxg Have added insurance logic if you want to do a quick rereview.

File: src/selectors/insurance.js [hunk 2]
```diff
@@ -17,6 +17,19 @@ export const selectInsurancePolicyDiscountRate = ({ insurancePolicy }) => {
...
+  const { patient: policyPatient, enteredBy: policyUser } = selectedInsurancePolicy ?? {};
+
+  const { isEditable: isPatientEditable } = policyPatient ?? {};
+
+  const { currentUser } = user ?? {};
+  const { id: currentUserId } = currentUser ?? {};
+  const { id: policyUserId } = policyUser ?? {};
+
+  return isPatientEditable || (currentUserId && policyUserId && currentUserId === policyUserId);
+};
```
Comment (joshxg) [inline src/selectors/insurance.js:31]: I'm confused why the user ID is being used here?

Comment (wlthomson) [reply src/selectors/insurance.js:31]: The use-case here is where you lookup a patient and add a new policy. That policy should remain editable, rather than the other policies, which should be uneditable. Does that make sense?

It's not a perfect solution; the assumption here is that if the policy was created by the current user, then the current user can edit it without causing sync issues (not a fool-proof solution, but the best compromise I can think of with existing data).

Comment (joshxg) [reply src/selectors/insurance.js:31]: ah I see - I imagined we would not allow creation of insurance policies for non-home store patients until there was a proper solution?

I suppose we never found a proper solution from this issue: https://github.com/sussol/msupply/issues/5491

Comment (joshxg) [reply src/selectors/insurance.js:31]: what is the behaviour now on desktop that you can have patients that aren't from the home store?

Comment (wlthomson) [reply src/selectors/insurance.js:31]: Yeah, it got a bit of a case of too many chefs! @mark-prins and I decided to allow creation of insurance policies but not editing of existing ones. I think you're right to lean on the conservative option, and that was our initial thinking too. The reason we came down on the side of allowing creation was that we considered it potentially a blocker to prescribing to that patient. 

I'm in the process of writing up docs for this, so this will be included. Have no problem with revisiting this decision!

Comment (joshxg) [reply src/selectors/insurance.js:31]: we can quite easily allow creation if we just add a `storeID` and make them a hybrid store data and patient data. This use of the user ID I think creates a worse outcome (for us, not the patient), where you have records editable in many stores?

Comment (wlthomson) [reply src/selectors/insurance.js:31]: > what is the behaviour now on desktop that you can have patients whose home store is another store?

Do you mean what is the insurance policy behaviour on desktop for non-local patients?

As I understand it, the patient API isn't implemented for desktop satellite sites, so you can't prescribe to non-local patients (as you can't seem them). My understanding is that the dispensing logic will be consistent with mobile, so the decision should probably be made here.

AFAIK, currently, you can't add insurance policies on desktop for non-local patients. However, as I understand it you also can't prescribe to them, so there's not much need.

Please correct me if any of the above is wrong! @mark-prins Have I missed anything?

Comment (joshxg) [reply src/selectors/insurance.js:31]: Was just checking as I thought I might have missed something and this was actually mimicing desktop behaviour (with the user stuff).

- Seems to be there are two worst case scenarios: A patient can't get their prescription as a user can't add an insurance policy for that patient or data inconsistencies due to being able to edit records in multiple stores.
- We maybe don't have time to fully solve these issues right now (?)

I think it is a lot easier to solve the former problem later (if we don't have time now), then having to deal and solve data issues (which IC have had a lot of, their data file is messsssed uuuup).

I'm sorry I'm happy to be told I'm wrong but I just think we need to make insurance policies uneditable and non-creatable(??) for non-local patients until there is a proper solution  ! obviously you don't, is it because the patient not getting a prescription worst case scenario is more important? It probably is.. 🤷‍♂

Comment (wlthomson) [reply src/selectors/insurance.js:31]: What you said is all correct, the reason for the workaround is to maintain pretty strong data consistency (the only case where a policy can be edited in two places is by the same user), while also avoiding the case where effectively a patient can't use their insurance because of constraints on foreign patients (if this occurs, the user is likely to create a duplicate patient and we will have data inconsistencies anyway).

So basically, it's an acknowledgement of both worst-cases you just mentioned, and an attempt to mitigate them both. Does that make sense?

**Comment by wlthomson** [comment]
@joshxg @mark-prins Going to go ahead and merge into the feature branch, but open to making changes and have created an issue for further discussion (#2680).

File: src/selectors/insurance.js (hunk 2 above)
Comment (mark-prins) [reply src/selectors/insurance.js:31]: sorry, bit late to the discussion. Ideally, should be locking down and preventing the addition of policies for 'foreign' patients. I think you've summarised the case well, Will. this solution isn't ideal but I think is the best of a bad bunch.
